File: PostsAndComments/app/src/main/java/com/example/postscommentsapp/presentation/viewmodel/CommentsViewModel.java
package com.example.postscommentsapp.presentation.viewmodel;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import java.util.List;
import java.util.concurrent.Executors;

import javax.inject.Inject;

import dagger.hilt.android.lifecycle.HiltViewModel;

@HiltViewModel
public class CommentsViewModel extends ViewModel {

    private final GetCommentsUseCase getCommentsUseCase;
    private final InsertCommentUseCase insertCommentUseCase;

    private final MutableLiveData<List<Comment>> commentsLiveData = new MutableLiveData<>();
    private final MutableLiveData<Boolean> loading = new MutableLiveData<>();
    private final MutableLiveData<String> error = new MutableLiveData<>();

    @Inject
    public CommentsViewModel(GetCommentsUseCase getCommentsUseCase,
                             InsertCommentUseCase insertCommentUseCase) {
        this.getCommentsUseCase = getCommentsUseCase;
        this.insertCommentUseCase = insertCommentUseCase;
    }

    public LiveData<List<Comment>> getComments() {
        return commentsLiveData;
    }

    public LiveData<Boolean> isLoading() { return loading; }

    public LiveData<String> getError() { return error; }

    public void loadComments(int postId) {
        loading.postValue(true);

        Executors.newSingleThreadExecutor().execute(() -> {
            try {
                List<Comment> comments = getCommentsUseCase.execute(postId);
                commentsLiveData.postValue(comments);
            } catch (Exception e) {
                error.postValue("Error al cargar comentarios.");
            } finally {
                loading.postValue(false);
            }
        });
    }

    public void addComment(Comment comment) {
        Executors.newSingleThreadExecutor().execute(() -> {
            try {
                insertCommentUseCase.execute(comment);
                loadComments(comment.getPostId());
            } catch (Exception e) {
                error.postValue("Error al insertar comentario.");
            }
        });
    }
}

File: PostsAndComments/app/src/main/java/com/example/postscommentsapp/presentation/viewmodel/CommentsViewModel.kt
package com.example.postscommentsapp.presentation.viewmodel

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.postscommentsapp.domain.model.Comment
import com.example.postscommentsapp.domain.usecase.GetCommentsUseCase
import com.example.postscommentsapp.domain.usecase.InsertCommentUseCase
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class CommentsViewModel @Inject constructor(
    private val getCommentsUseCase: GetCommentsUseCase,
    private val insertCommentUseCase: InsertCommentUseCase
) : ViewModel() {

    val comments = MutableLiveData<List<Comment>>()
    val loading = MutableLiveData<Boolean>()
    val error = MutableLiveData<String?>()

    fun loadComments(postId: Int) {
        viewModelScope.launch {
            loading.value = true
            try {
                comments.value = getCommentsUseCase.execute(postId)
            } catch (e: Exception) {
                error.value = "Error cargando comentarios"
            }
            loading.value = false
        }
    }

    fun addComment(postId: Int, text: String) {
        viewModelScope.launch {
            try {
                insertCommentUseCase.execute(
                    Comment(
                        id = 0,
                        postId = postId,
                        comment = text,
                        createdAt = System.currentTimeMillis()
                    )
                )
                loadComments(postId)
            } catch (e: Exception) {
                error.value = "No se pudo insertar el comentario"
            }
        }
    }
}
